backend/src/chat.gateway.ts

```diff
@@ -0,0 +1,144 @@
+import { TokenUserDto } from './auth/dto/token-user.dto';
+import { plainToInstance } from 'class-transformer';
+import { ResponseMessageDto } from './messages/dto/response-message.dto';
+import { MembershipsService } from './memberships/memberships.service';
+
+export class ChatJoinDto {
+  channel: string;
+}
+
+export class ChatSendDto {
+  channel: string;
+  message: string;
+}
+
+@WebSocketGateway({ cors: true })
+export class ChatGateway
+  implements OnGatewayInit, OnGatewayConnection, OnGatewayDisconnect
+{
+  constructor(
+    private readonly messagesService: MessagesService,
+    private readonly usersService: UsersService,
+    private readonly membershipsService: MembershipsService,
+  ) {}
+
+  @WebSocketServer() server: Server;
+  private logger: Logger = new Logger('ChatGateway');
+  private authenticatedSockets: Map<string, User> = new Map();
+
+  private checkAuth(client: Socket): boolean {
+    if (this.authenticatedSockets.has(client.id)) return true;
+    throw new WsException('Not Authorized');
+  }
+
+  @SubscribeMessage('chat-join')
+  handleJoin(client: Socket, payload: ChatJoinDto) {
+    this.logger.log(`${client.id} wants to join room [${payload.channel}]`);
+    this.checkAuth(client);
+    // Check if User has permission to join channel here
+    client.join(payload.channel);
+    return `Successfully joined channel ${payload.channel}`;
+  }
+
+  @SubscribeMessage('chat-leave')
+  async handleLeave(client: Socket, payload: ChatJoinDto) {
+    this.logger.log(`${client.id} wants to leave room [${payload.channel}]`);
+    this.checkAuth(client);
+    // Check if User leaving affects ownership
+    client.leave(payload.channel);
+    const membershipArray = await this.membershipsService.findAll({
+      channel: payload.channel,
+      user: this.authenticatedSockets.get(client.id).id.toString(),
+    });
+    if (membershipArray && membershipArray.length === 1) {
+      const membership = membershipArray[0];
+      if (
+        membership &&
+        membership.id &&
+        membership.userId === this.authenticatedSockets.get(client.id).id &&
+        membership.channelId === parseInt(payload.channel)
+      ) {
+        await this.membershipsService.remove(membership.id);
+      }
+    }
+    return `Successfully left channel ${payload.channel}`;
+  }
+
+  @SubscribeMessage('chat-auth')
+  async handleAuth(client: Socket, payload: any) {
+    this.logger.log(
+      `Client ${client.id} is trying to auth socket with token: ${payload.authorization}`,
+    );
+    try {
+      const token = payload.authorization;
+      const decode = jwt.verify(token, process.env.JWT_SECRET_KEY);
+      const user = await this.usersService.findOne(
+        +(decode as TokenUserDto).id,
+      );
+      if (!user) {
+        throw new WsException('Invalid Token');
+      }
+      this.authenticatedSockets.set(client.id, user);
+      return 'SUCCESS';
+    } catch (err) {
+      this.authenticatedSockets.delete(client.id);
```

Comment:
The reasoning behind this was just in case an authenticated socket tried to authenticate with a refreshed token or as another user, and it failed, then we would be sure that there was no entry leftover in the authenticatedSockets. But I can see that it might be unnecessary.

If you think it is better without, then I'll remove it

Reply:
Yes, you are right, particularly in the case of a refreshed token.